File: heron/api/src/java/com/twitter/heron/streamlet/WindowConfig.java
//  Copyright 2017 Twitter. All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

package com.twitter.heron.streamlet;


import java.time.Duration;

import com.twitter.heron.streamlet.impl.WindowConfigImpl;

/**
 * WindowConfig allows Streamlet API users to program window configuration for operations
 * that rely on windowing. Currently we only support time/count based
 * sliding/tumbling windows.
 */
public interface WindowConfig {
  /**
   * Creates a time based tumbling window of windowDuration
   * @param windowDuration the duration of the tumbling window
   * @return WindowConfig that can be passed to the transformation
   */
  static WindowConfig TumblingTimeWindow(Duration windowDuration) {
    return new WindowConfigImpl(windowDuration, windowDuration);
  }

  /**
   * Creates a time based sliding window with windowDuration as the window duration
   * and slideInterval as slideInterval
   * @param windowDuration The Sliding Window duration
   * @param slideInterval The sliding duration
   * @return WindowConfig that can be passed to the transformation
   */
  static WindowConfig SlidingTimeWindow(Duration windowDuration, Duration slideInterval) {
    return new WindowConfigImpl(windowDuration, slideInterval);
  }

  /**
   * Creates a count based tumbling window of size windowSize
   * @param windowSize the size of the tumbling window
   * @return WindowConfig that can be passed to the transformation
   */
  static WindowConfig TumblingCountWindow(int windowSize) {
    return new WindowConfigImpl(windowSize, windowSize);
  }

  /**
   * Creates a count based sliding window with windowSize as the window countsize
   * and slideSize as slide size
   * @param windowSize The Window Count Size
   * @param slideSize The slide size
   * @return WindowConfig that can be passed to the transformation
   */
  static WindowConfig SlidingCountWindow(int windowSize, int slideSize) {
    return new WindowConfigImpl(windowSize, slideSize);
  }
}

File: heronpy/streamlet/windowconfig.py
# Copyright 2016 - Twitter, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
'''windowconfig.py: module for defining windowconfig'''

import datetime

class WindowConfig(object):
  """WindowConfig allows streamlet API users to program window configuration for operations
     that rely on windowing. Currently we only support time/count based
     sliding/tumbling windows.
  """
  def __init__(self, window_duration, slide_interval):
    self._window_duration = window_duration
    self._slide_interval = slide_interval
    if self._window_duration.microseconds > 0 or self._slide_interval.microseconds > 0:
      raise RuntimeError("Python Windowing curently only supports second resolution")

  @staticmethod
  def create_tumbling_window(window_duration):
    return WindowConfig(window_duration, window_duration)

  @staticmethod
  def create_sliding_window(window_duration, slide_interval):
    if isinstance(window_duration, int):
      window_duration = datetime.timedelta(seconds=window_duration)
    if isinstance(slide_interval, int):
      slide_interval = datetime.timedelta(seconds=slide_interval)
    if not isinstance(window_duration, datetime.timedelta):
      raise RuntimeError("Window Duration has to be of type datetime.timedelta")
    if not isinstance(slide_interval, datetime.timedelta):
      raise RuntimeError("Slide Interval has to be of type datetime.timedelta")
    return WindowConfig(window_duration, slide_interval)
